src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -27,8 +25,50 @@ public ChangedText(SourceText oldText, SourceText newText, ImmutableArray<TextCh
             Debug.Assert(!changeRanges.IsDefault);
 
             _newText = newText;
-            _weakOldText = new WeakReference<SourceText>(oldText);
-            _changes = changeRanges;
+            _info = new ChangeInfo(changeRanges, new WeakReference<SourceText>(oldText), (oldText as ChangedText)?._info);
+        }
+
+        private class ChangeInfo
+        {
+            public ImmutableArray<TextChangeRange> ChangeRanges { get; }
+
+            // store old text weakly so we don't form unwanted chains of old texts (especially chains of ChangedTexts)
+            // used to identify the changes in GetChangeRanges.
+            public WeakReference<SourceText> WeakOldText { get; }
+
+            public ChangeInfo Previous { get; private set; }
+
+            public ChangeInfo(ImmutableArray<TextChangeRange> changeRanges, WeakReference<SourceText> weakOldText, ChangeInfo previous)
+            {
+                this.ChangeRanges = changeRanges;
+                this.WeakOldText = weakOldText;
+                this.Previous = previous;
+                Clean();
+            }
+
+            // clean up 
+            private void Clean()
+            {
+                // look for last info in the chain that still has reference to old text
+                ChangeInfo lastInfo = this;
+                for (var info = this; info != null; info = info.Previous)
```

Comment:
My thought is:
- You can be keeping unbounded trail of ChangeInfos if one very old text made it into Gen2 and now will not be collected for a while. 
Not only you will be keeping that trail, but you will continue  visiting the whole list all the way over and over to find that "stuck" text is still alive. It seems to be the cost of walks could be N^2 where N is the length of the trail, so may become an expense by itself.

- even the trail of 10000 ChangeInfos happens to be real reachable and you are asked to compute diffs between N0 and N10000, do you want to do it by merging?

It feels like having a limit on how long the chain would make it harder to get into degraded states while not penalizing common scenarios.